tests/common/exceptions.h

```diff
@@ -0,0 +1,302 @@
+      // There should be no compilation failures, but every CHECK_THROWS_MATCHES
+      // should fail with this matcher
+      return false;
+    } else {
+      return other.category() == m_category;
+    }
+  }
+
+  std::string describe() const override {
+    return std::string("has category: '") + m_category.name() + "'";
+  }
+
+ private:
+  const std::error_category& m_category;
+};
+
+/**
+ *  Matcher for sycl::errc error codes
+ *  C++ provides semantic match for std::error_code by operator==
+ */
+struct matcher_exception_code_semantic_match
+    : public Catch::Matchers::MatcherGenericBase {
+  matcher_exception_code_semantic_match(const sycl::errc& code_value)
+      : m_code_value(code_value) {}
+
+  static constexpr bool is_supported_by_implementation() {
+    return detail::has_member::code<sycl::exception>::value;
+  }
+
+  bool match(const sycl::exception& other) const {
+    if constexpr (!is_supported_by_implementation()) {
+      // There should be no compilation failures, but every CHECK_THROWS_MATCHES
+      // should fail with this matcher
+      return false;
+    } else {
+      // Semantic match
+      return other.code() == m_code_value;
+    }
+  }
+
+  std::string describe() const override {
+    std::ostringstream result;
+    result << "has code semantically matching to ";
+    result << to_integral(m_code_value);
+    if constexpr (detail::has_function::sycl_make_error_code<>::value) {
+      result << " (" << sycl::make_error_code(m_code_value).message() << ")";
+    }
+    return result.str();
+  }
+
+ private:
+  const sycl::errc& m_code_value;
+};
+
+/**
+ *  Matcher for backend-specific category and error codes
+ */
+template <sycl::backend Backend, typename CodeT>
+class matcher_equals_exception_for
+    : public Catch::Matchers::MatcherGenericBase {
+ public:
+  matcher_equals_exception_for(const CodeT& code_value)
+      : m_code_value(code_value) {}
+
+  static constexpr bool is_supported_by_implementation() {
+    return detail::has_member::code<sycl::exception>::value &&
+           detail::has_member::category<sycl::exception>::value;
+  }
+
+  bool match(const sycl::exception& other) const {
+    if constexpr (!is_supported_by_implementation()) {
+      // There should be no compilation failures, but every CHECK_THROWS_MATCHES
+      // should fail with this matcher
+      return false;
+    } else {
+      // Error code is validated by value, with no semantic match
+      return (other.category() == expected_category()) &&
+             (other.code().value() == m_code_value);
+    }
+  }
+
+  std::string describe() const override {
+    std::ostringstream result;
+    result << "has code value " << to_integral(m_code_value);
+    result << " for backend-specific category '" << expected_category().name();
+    result << "'";
+    return result.str();
+  }
+
+ private:
+  const std::error_category& expected_category() const {
+    return sycl::error_category_for<Backend>();
```

Comment:
Seems like this is not supported by any (even DPC++, for me?) implementation yet.

Reply:
Also under feature flag now )